# Linear Regression: The Mother of Machine Learning Algorithms

A **function** is a type of relation in which an input always has a fixed output.
Mathematically, if we have x1 and x2, then:

f(x1) = f(x2)  if x1 = x2

One common type of function we often use is the **linear function**, which has the form:

f(x) = m * x + b

With this function, many things become easier. Whenever we have an input x, we can calculate the output.

**Example:**

f(x) = 3 * x + 5

If x = 2, then:

f(2) = 3 * 2 + 5 = 11

Now, here’s an interesting question:
If we give a machine a bunch of input-output pairs, will the machine be able to predict the values of m and b such that it can represent the linear function?
This is exactly what **Linear Regression** does!


In [13]:
# lets generate input and out put for the above function f(x) = 3x+5
input_set = [[i] for i  in range(10)] # here is a  we are using for input
output_set = [3*x+5 for x in range(10)] # here is output of the input 3*input +5
print(input_set)
print(output_set)


[[0], [1], [2], [3], [4], [5], [6], [7], [8], [9]]
[5, 8, 11, 14, 17, 20, 23, 26, 29, 32]


#  training the model

In [17]:
from sklearn import linear_model

model = linear_model.LinearRegression()
model.fit(input_set,output_set) # this make the model learn from input and out put

print(f"the equation of the line is f(x) = {model.coef_[0]}x+{model.intercept_}")


the equation of the line is f(x) = 3.000000000000001x+4.9999999999999964


# The model almost computed the equation as folows
the equation of the line is f(x) = 3.000000000000001x+4.9999999999999964

The model is able to learn from the data and predict the function, but not all relationships are perfect, and sometimes we provide it with complex data.

# house price prediction

In [21]:
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, r2_score
import matplotlib.pyplot as plt

# 2. Example real dataset (house size vs price)
X = [[50], [60], [70], [80], [90], [100], [110], [120], [130], [140]]  # size in square meters
y = [150, 180, 210, 240, 270, 300, 330, 360, 390, 420]                 # price in thousands

X_train, X_test , y_train, y_test = train_test_split(X,y, test_size = 0.2,random_state = 42)
model = LinearRegression()

model.fit(X_train,y_train)

y_pred = model.predict(X_test)

print(f"for input {X_test}: the model prediction is {y_pred}")

# graphically
plt.scatter()



